### PR TITLE
Docs/guide

### PR DESCRIPTION
**Pull request by ashdtu**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#1103 : Better documentation of Guide example.
#1153: Improve Tensor introduction
#1154: Improve Dataset transform guide

### Changes
Corresponding to each issue, here are the changes in this PR:
#1103 : The changes suggested bridge the gap between the example code provided in examples/guide vs what is written in guide. The documentation currently lacked the partitioning of code into model, data, training files similar to the example code. This sometimes made it hard to follow if for eg someone is copy-pasting code from there. Fixed these issues and added helpful comments with reference to original example.

#1153
* Added introduction of  Tensor with rank, similarity/diff for people coming from PyTorch.
* Initialization from data, floats api. .convert() , elem etc.
*  Underline importance of ownership and cloning with example.
* No stateful operations : **TODO** (discuss)

#1154 
* Added introduction of dataset transform with example usages.
* Batcher omitted as already present in current guide in mnist tutorial and other concepts covered by Rust introduction commit.

### CODE REVIEW

File: burn-book/src/building-blocks/dataset.md [hunk 1]
```diff
@@ -32,6 +32,50 @@ distributions.
...
+the raw dataset into train/test splits. Can be initialized with a seed to ensure reproducibility. 
+```rust, ignore
+let dataset = ShuffledDataset<DbPedia, DbPediaItem>::with_seed(dataset, 42);
+```
+
+* **PartialDataset**: This transform is useful to return a view of the dataset with a specified start, end index. Used
+to create train/val/test splits. Below example we show how to chain ShuffledDataset and PartialDataset to create splits.
+```rust, ignore
+// define chained dataset type here for brevity
+let PartialData = PartialDataset<ShuffledDataset<DbPedia, DbPediaItem>>;
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/dataset.md:61]: Prefer using `type` for type alias

```rust
type PartialData = PartialDataset<ShuffledDataset<DbPedia, DbPediaItem>>;
```

File: burn-book/src/building-blocks/dataset.md [hunk 2]
```diff
@@ -32,6 +32,50 @@ distributions.
 | `MapperDataset`   | Computes a transformation lazily on the input dataset.                                                                   |
 | `ComposedDataset` | Composes multiple datasets together to create a larger one without copying any data.                                     |
 
+Let us look at the basic usages of each dataset transform and how they can be composed together. The full documentation
+of each transform can be found at the [API reference](https://burn.dev/docs/burn/data/dataset/transform/index.html).
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/dataset.md:36]: I would add a sentence or two specifying that the dataset transformation are lazy by default except when specified otherwise.

File: burn-book/src/building-blocks/tensor.md [hunk 3]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
 Note that the specific element types used for `Float`, `Int`, and `Bool` tensors are defined by
 backend implementations.
 
-## Operations
+In contrast to PyTorch where a Tensor is characterised by the size of it's each dimension, a Burn Tensor is
+characterized by the number of dimensions D in its declaration. The actual shape of the tensor is inferred from
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:17]: I don't think PyTorch is characterized by the size of its each dimension. There is nothing specified for the element type, the shape, or the rank. All of those are runtime properties. In contrast, in Burn, we have those specified as types!

File: burn-book/src/building-blocks/tensor.md [hunk 4]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+let tensor_1 = Tensor::<Backend, 1>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0]);
+
+// incorrect: let tensor_1 = Tensor::<Backend, 5>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0]);
+// this will lead to an error and is for creating a 5-D tensor
+```
+
+### Initialization
+
+Burn Tensors are primarily initialized using the `from_data()` method which takes the `Data` struct as input. 
+The `Data` struct has two fields: value & shape. To access the data of any tensor, we use the method `.to_data()`.
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:30]: `to_data()` if we want to reuse the tensor afterward, or `into_data()` otherwise.

File: burn-book/src/building-blocks/tensor.md [hunk 5]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+// correct: Tensor is 1-Dimensional with 5 elements
+let tensor_1 = Tensor::<Backend, 1>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0]);
+
+// incorrect: let tensor_1 = Tensor::<Backend, 5>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0]);
+// this will lead to an error and is for creating a 5-D tensor
+```
+
+### Initialization
+
+Burn Tensors are primarily initialized using the `from_data()` method which takes the `Data` struct as input.
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:29]: Also add the fact that `Data` can be built from `slices`, `vectors`, etc. See https://github.com/tracel-ai/burn/blob/main/burn-tensor/src/tensor/data.rs

File: burn-book/src/building-blocks/tensor.md [hunk 6]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+Burn Tensors are primarily initialized using the `from_data()` method which takes the `Data` struct as input. 
+The `Data` struct has two fields: value & shape. To access the data of any tensor, we use the method `.to_data()`.
+Let's look at a couple of examples for initializing a tensor from different inputs. 
+
+```rust, ignore
+// Initialization from a float array
+let tensor_1 = Tensor::<Backend, 1>::from_data([1.0, 2.0, 3.0]);
+
+// equivalent to (recommended)
+let tensor_2 = Tensor::<Backend, 1>::from_floats([1.0, 2.0, 3.0]);
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:38]: The different with from `floats` is that the element type is f32, where from data the element type is defined by the backend. When the backend is generic, you have to use some sort of convertion.

File: burn-book/src/building-blocks/tensor.md [hunk 7]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+        age: 10,
+        height: 100,
+        weight: 50.0
+    };
+let tensor_3 = Tensor::<Backend, 1>::from_data(Data::from([bmi.age as f32, bmi.height as f32, bmi.weight]).convert());
+
+```
+The `.convert()` method for Data struct is called in the last example to ensure that the data's primitive type is 
+consistent across all backends. The convert operation can also be done at element wise level as: 
+`let tensor_4 = Tensor::<B, 1, Int>::from_data(Data::from([(item.age as i64).elem()])`.
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:58]: Specify that using `elem` is only possible when the trait `ElementConversion` is imported.

File: burn-book/src/building-blocks/tensor.md [hunk 8]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+    println!("{:?}", input.to_data());      // Success: [0.0, 0.33333334, 0.6666667, 1.0]
+    
+    // Notice that max, min have been moved in last operation so the below print will give an error.
+    // If we want to utlize them as well for further operations, they will need to be cloned in similar fashion.
+    // println!("{:?}", min.to_data());    
+```
+
+We don't need to be worried about memory overhead because with cloning, the tensor's buffer isn't copied, 
+and only a reference to it is increased. This makes it possible to determine exactly how many times a tensor is used, 
+which is very convenient for reusing tensor buffers and improving performance. For that reason, we don't provide
```
Comment (nathanielsimard) [inline burn-book/src/building-blocks/tensor.md:91]: for reusing tensor buffers or even fusing operations into a single kernel (link to burn fusion maybe).

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Some small corrections, see my comments bellow.

File: burn-book/src/building-blocks/tensor.md [hunk 9]
```diff
@@ -13,14 +13,85 @@ Tensor<B, D, Bool>     // Bool tensor
...
+
+```rust, ignore
+    let input = Tensor::<Wgpu, 1>::from_floats([1.0, 2.0, 3.0, 4.0]);
+    let min = input.clone().min();
+    let max = input.clone().max();
+    let input = (input.clone() - min.clone()).div(max - min);
+    println!("{:?}", input.to_data());      // Success: [0.0, 0.33333334, 0.6666667, 1.0]
+    
+    // Notice that max, min have been moved in last operation so the below print will give an error.
+    // If we want to utlize them as well for further operations, they will need to be cloned in similar fashion.
```
Comment (louisfd) [inline burn-book/src/building-blocks/tensor.md:85]: There's a typo in utilize, i prefer `use` to utilize in general

**Comment by antimora** [comment]
Quick feedback re formatting (haven't read the content in details yet). 

I recommend you format the text using something like prettier if you have a VS/Neovim plug-in. We have a configuration for prettier: https://github.com/tracel-ai/burn/blob/main/burn-book/.prettierrc.json (or you can define yours)

This helps formatting tables and wrapping markdown text.

**Comment by ashdtu** [comment]
Many thanks @nathanielsimard, @louisfd, and @antimora for the suggestions. Incorporated in the latest commit. Let me know if it's good to go.

**Review by nathanielsimard — APPROVED** [review]

**Comment by ashdtu** [comment]
Thank again for the review. Made the minor fixes suggested. Merging now!
